command/validate.py

```diff
@@ -1,8 +1,243 @@
+
+            # check certified value's type in particular. should be string, not bool
+            if "certified" not in annotations:
+                logger.warning("csv metadata.annotations.certified not defined.")
+            else:
+                isString = isinstance(annotations["certified"], str)
+                if not isString:
+                    logger.error("metadata.annotations.certified is not of type string")
+                    valid = False
+
+            # if alm-examples is defined, check that its value is valid json
+            if "alm-examples" in annotations:
+                try:
+                    valid_json = json.loads(annotations["alm-examples"])
+                except:
+                    logger.error("metadata.annotations.alm-examples contains invalid json string")
+                    valid = False
+
+        else:
+            logger.warning("csv metadata.annotations not defined.")
+
+        return valid
+
+    def _pkgs_validation(self, bundleData):
+        valid = True
+        logger.info("Validating packages.")
+
+        pkgs = bundleData[self.pkgsKey]
+
+        for pkg in pkgs:
+            if "packageName" in pkg:
+                logger.info("Evaluating package %s", pkg["packageName"])
+            else:
+                logger.error("packageName not defined.")
+                valid = False
+
+            if "channels" in pkg:
+                channels = pkg["channels"]
+                if len(channels) == 0:
+                    logger.error("no package channels defined.")
+                    valid = False
+                else:
+                    csvNames = []
+                    for csv in bundleData[self.csvKey]:
+                        try:
+                            csvName = csv["metadata"]["name"]
+                            csvNames.append(csvName)
+                        except:
+                            pass
+                    for channel in channels:
+                        if "name" not in channel:
+                            logger.error("package channel.name not defined.")
+                            valid = False
+
+                        if "currentCSV" not in channel:
+                            logger.error("package channel.currentCSV not defined.")
+                        else:
+                            if channel["currentCSV"] not in csvNames:
+                                logger.error("channel.currentCSV %s is not included in list of csvs")
+                                valid = False
+
+            else:
+                logger.error("package channels not defined.")
+                valid = False
+
+        return valid
+
+    def _type_validation(self, bundleData, typeName, validator):
+        valid = True
+        if typeName in bundleData:
+            if len(bundleData[typeName]) != 0:
+                valid = validator(bundleData)
+            else:
+                logger.error("Bundle does not contain any %s." % typeName)
```

Comment:
It's possible for a valid bundle to be created without any CRDs if the CSV doesn't list any. An example is the Service Catalog operator: https://github.com/operator-framework/community-operators/blob/master/community-operators/bundles/svcat/bundle.0.1.34.yaml#L7

Reply:
Addressed in latest commit.